### ISSUE TITLE
external vs view into shared buffer?

### ISSUE THREAD

**Issue by cyrush** [LLNL/axom#45]
@nselliott pointed out an issue with reading external buffers in the VisIt Blueprint plugin. 

The files were created with `src/axom/sidre/examples/sidre_createdatastore.cpp`
I found the issue with the VisIt plugin, however I am still a bit confused b/c I did not expect these fields to be written out as "external".

Here is the snippet that creates the views:

```
  // "x", "y", and "z" are three views into a shared Sidre buffer object that
  // holds 3 * nodecount doubles.  These views might describe the location of
  // each node in a 16 x 16 x 16 hexahedron mesh.  Each view is described by
  // number of elements, offset, and stride into that data.
  Buffer* buff = ds->createBuffer(sidre::DOUBLE_ID, 3*nodecount)->allocate();
  nodes->createView("x", buff)->apply(sidre::DOUBLE_ID, nodecount, 0, 3);
  nodes->createView("y", buff)->apply(sidre::DOUBLE_ID, nodecount, 1, 3);
  nodes->createView("z", buff)->apply(sidre::DOUBLE_ID, nodecount, 2, 3);
  // _first_example_nodes_end
```

(With the fix) in VisIt I am definitely reading them as external views from the HDF5 file:

```
loading mesh/coordsets/coords/values/x as sidre external view
relay:io::hdf5_read domain 0 : datagroup_0000000/sidre/external/mesh/coordsets/coords/values/x
```

Is this expected? I expected them to simply be a view into a known buffer. 

@kennyweiss @agcapps any thoughts?

**Comment by cyrush** [comment]
I think the issue is actually here:

```
void setup_blueprint_coords(DataStore* ds, Group* coords)
{
  // _blueprint_restructure_coords_start
  // Set up the coordinates as Mesh Blueprint requires
  coords->createViewString("type", "explicit");
  // We use prior knowledge of the layout of the original datastore
  View* origv = ds->getRoot()->getView("nodes/xs");
  Group* conduitval = coords->createGroup("values");
  conduitval->createView("x", sidre::DOUBLE_ID,
                         origv->getNumElements(),
                         static_cast<double*>(origv->getArray()));
  origv = ds->getRoot()->getView("nodes/ys");
  conduitval->createView("y", sidre::DOUBLE_ID,
                         origv->getNumElements(),
                         static_cast<double*>(origv->getArray()));
  origv = ds->getRoot()->getView("nodes/zs");
  conduitval->createView("z", sidre::DOUBLE_ID,
                         origv->getNumElements(),
                         static_cast<double*>(origv->getArray()));
  // _blueprint_restructure_coords_end
}
```
We are severing the relationship to the view.  This means that the data is actually being written twice (once for the actual buffer and again here as an external buffer)

**Comment by kennyweiss** [comment]
@cyrush -- so, it sounds like the problem is that we should be using this function:
https://github.com/LLNL/axom/blob/9017f6c2ff279d7a4c4037783311ba70dcd6af33/src/axom/sidre/core/Group.hpp#L505-L524

rather than this one:
https://github.com/LLNL/axom/blob/9017f6c2ff279d7a4c4037783311ba70dcd6af33/src/axom/sidre/core/Group.hpp#L598-L618

--- 

Note: this would also make the code cleaner.  I.e. this call
```cpp
  conduitval->createView("x", sidre::DOUBLE_ID,
                         origv->getNumElements(),
                         static_cast<double*>(origv->getArray()) );
```
can be replaced by this one:
```cpp
  conduitval->createView("x", sidre::DOUBLE_ID,
                         origv->getNumElements(),
                         origv->getBuffer() );
```

**Comment by nselliott** [comment]
@cyrush The situation here is that there are two DataStores--The first allocates and initializes the data using regular Sidre buffers, and the second creates Views that point at that data as raw pointers to external data.  Only the second is used in the I/O in this example.

@kennyweiss It depends on what we're trying to test.  As it is now, the first DataStore simulates an application that owns its data and could have set it up in any manner, while the second DataStore is created to give a Sidre description of that existing data, which it only sees as raw pointers.  The fact that Sidre was used to create the first doesn't matter to the second.  Your proposed change would test a case where the second DataStore is re-using existing Sidre buffers that it knows about from the first DataStore.

**Comment by cyrush** [comment]
@nselliott Thanks for the details.  I did not fully understand the example. For testing its very useful -- it show some complex stuff that we support. 

The concern I have is that the example implies we have to create a new datastore in order to output Blueprint meshes. That's not the case and that's not how I would recommend using it, yet thats our only example in the docs :-) We should think about a simpler example to help teach people for the docs.

**Comment by nselliott** [comment]
@cyrush I should revisit this to get simpler example in the docs without losing this more complicated test coverage.

**Comment by cyrush** [comment]
@nselliott sounds like a good plan

**Comment by nselliott** [comment]
@cyrush @kennyweiss Am I correct to observe that it is unsupported behavior for a View created in one DataStore to point to an existing Buffer in another DataStore?  (Using explicit Buffer syntax as in comment from @kennyweiss on May 23.)  It doesn't trigger an error, but if DataStore A creates a Buffer, and then DataStore B creates a View that points to that Buffer, it looks like whichever DataStore is deallocated first deallocates the buffer without the other DataStore knowing.  I haven't fully diagnosed if this is actually happening, but it seems likely it is what is causing a crash to occur as I'm trying to modify this test.  Seems like either this usage should trigger an error unless we want to find a way to truly support it.

I can do my test mods using only one DataStore, so this isn't blocking me from anything.

**Comment by cyrush** [comment]
@nselliott correct,  i don't think we want to support views that into buffers in another datastore, except for the fully "external" view case.  Does that make sense?

External views should be ok, however they come with the same caveats -- you as the user have to make sure the pointer is valid. If the own datastore cleans up, you can't use that pointer anymore.

I do think it would be good to protected agains some of these cases. That said, I really don't think we should recommend folks using multiple data stores.  That should only be needed for super advanced use cases, and I'd argue folks would need as strong handle on Sidre implementation details.

**Comment by agcapps** [comment]
When I wrote that example, I was trying to show how to export some data stored in Sidre according to the blueprint.  The only way I could think to do this was to create a new blueprint-conforming Datastore that pointed into the original Datastore, and save the conforming Datastore.  It is not simple and does make race conditions.

@cyrush and @nselliott , it makes sense to me that a View should not be constructed with a Buffer from another Datastore.  If a View actually does need to point into data residing in another Datastore, the user should make it an external View (basically what you said, @cyrush ).

**Comment by nselliott** [comment]
@agcapps I've been able to modify things so that everything exists in one DataStore, which allows us to use Buffers, and I think it makes things easier to explain, but unfortunately this makes the documentation you wrote obsolete, including the PNG figures.  I can edit the text, but the figures will be more difficult.

**Comment by agcapps** [comment]
No problem: when you have a branch or PR, point me to it and I'll update the figures.
